### ISSUE TITLE
Variant propagation: allow non-root variants

### ISSUE THREAD

**Issue by adamjstewart** [LLNL/spack#34302]
### Summary

#29761 introduced variant propagation (++, --, ==). I would like this feature to work even when the root spec does not have this variant.

### Rationale

Imagine a situation in which package A depends on B and C, which depend on D and E. You would like to build all of the packages in your DAG using `+mpi`, but package A does not have an MPI variant. Currently, this requires you to know which packages have this variant and which don't:
```console
$ spack install A ^B++mpi ^C++mpi
```
It would be great if we could instead use:
```console
$ spack install A++mpi
```
#32893 provides another example where this feature is needed. In CI, I would like to be able to build the cross product of the following:

* all ML packages
* the following variants:
  * `+cuda ~rocm cuda_arch=...`
  * `+rocm ~cuda amdgpu_target=...`
  * `~cuda ~rocm`

At the moment, there's no way to do this other than by having separate pipelines for all of these specs. This causes us to rebuild 90% of packages twice just because they can't be shared across pipelines. Once this feature is implemented, then the following would be possible:
```yaml
spack:
  specs:
  - matrix:
    - [py-torch, py-torchvision, ...]
    - ["++cuda ~~rocm cuda_arch==...", "++rocm ~~cuda amdgpu_target==...", "~~cuda ~~rocm"]
```

Another use case is support for `requires:` in `packages.yaml`. At the moment, `requires:` doesn't work for variants that don't exist in all packages. If we allow variant propagation to work for variants that don't exist in the root spec, then the following would be possible:
```yaml
packages:
  all:
    require: ~~mpi
```

### Description

I don't know the details of the implementation in #29761, but I'm hoping it won't be hard to allow these kinds of variants even if the root spec does not support them.

Note that the following should also work without concretization error:
```console
$ spack install zlib ++nonexistent_variant
```
Without this I can't properly support the above CI use case.

### Additional information

I've already spoken to @becker33 about this feature request, @RikkiButler20 may also be interested in this.

I think I complained about the limitations of `requires:` to @alalazo or @scheibelp, so this feature request would also improve that situation too.

### General information

- [X] I have run `spack --version` and reported the version of Spack
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by adamjstewart** [comment]
This was fixed/implemented at some point, @alalazo may know exactly when/where.

**Comment by adamjstewart** [comment]
Only `requires` was fixed, the rest of the features requested in this issue are implemented in #42931.
